File: 198.house-robber.py
#
# @lc app=leetcode id=198 lang=python3
#
# [198] House Robber
#

# @lc code=start
class Solution:
    def rob(self, nums: List[int]) -> int:
        # maximum value of robbing[i] would be derived from dp[i+2] & dp[i+3] (no dp[i+4] since it's always lower than dp[i+2])
        n = len(nums)
        dp =[0]*n
        for i in range(n-1, -1, -1):
            dp[i] = nums[i]+(max(dp[i+2] if i+2<n else 0, dp[i+3] if i+3<n else 0))
        return max(dp[0], dp[1] if n>1 else -1)

# @lc code=end

File: 198.house-robber.go
/*
 * @lc app=leetcode id=198 lang=golang
 *
 * [198] House Robber
 */

// @lc code=start
func rob(nums []int) int {
	// each house i, the best answer is either rob or not rob
	// (the value of this house + answer from i-2) or (answer from i-1)
	n := len(nums)
	ans := make([]int, n)
	for i, val := range nums {
		if i == 0 {
			ans[i] = val
			continue
		}
		prevAns := 0
		if i-2 >= 0 {
			prevAns = ans[i-2]
		}
		ans[i] = max(prevAns+val, ans[i-1])
	}
	return ans[n-1]
}

// @lc code=end

